README.md
# ux-genie

This repo is deprecated. The ux-genie module is now in the [genie repo](https://github.com/kentcdodds/genie/tree/master/src/lamps/angular).

CHANGELOG.md
# Latest

 - Deprecated `ux-genie` as a standalone repo/component. It now comes with [genie](http://github.com/kentcdodds/genie).

# 0.3.3

 - Forgot to uglify to uxGenie.min.js - Should really probably just make a build for this...

# 0.3.2

 - `ux-genie` - Adding commonjs style module support
 - `ux-lamp` - No longer calculating math when in subcontext
 - `ux-lamp` - Improved document click handler to determine whether to toggle visibility
 - `ux-lamp` - Improved lampVisible watcher to only focus on the input if necessary.

# 0.3.1

 - `ux-lamp` - Changing storage options. The `local-storage` attribute's value is now used as the key for the local storage value.
 - `ux-lamp` - New attribute `firebase` now allows you to store your data in a firebase. See README for more info.

# 0.3.0

 - `genie-wish`
   - removed the wish-action attribute so it no longer has an isolate scope.
 - `ux-lamp`
   - Added ability to press "tab" on sub context wishes.
   - Altered the template a bit to give more CSS control.
   - Improved scope.$apply stuff to be safe. Fixed subcontext wish execution.
   - Not allowing genie to be visible when genie is disabled.
   - Adding custom template option. Also changing the default template slightly which is the reason for the major version bump.

# 0.2.4

 - `genie-wish` - made it so you can escape commas for magic words like so: "People\, places\, and things".
 - `ux-lamp` - added subContext capabilities (see README).
 - `ux-lamp` - added simple calculator.
 - `ux-lamp` - fixed a couple of bugs...

# 0.2.3

 - `ux-lamp` bug fix. Not including the `lamp-visible`, `wish-callback`,
 or `local-storage` caused angular to crash with [this error](http://docs.angularjs.org/error/$compile:nonassign?p0=undefined&p1=uxLamp)

# 0.2.2

 - `genie-wish` directive will now look at the content of the element
 for the wish's magic word if there is no value for the `genie-wish`,
 `name`, or `id` attributes.
 - `genie-wish` now has a new attribute called `ignore-wish` which
 the directive observes for changes. If it is set to true during the
 link phase, the wish will not be registered until the attribute is
 changed to "true".
 - This change log :)

test/unit/directiveSpec.js
'use strict';

/* jasmine specs for directives go here */

describe('ngGenie', function() {
  beforeEach(module('ngGenie.directive'));

  describe('ng-genie', function() {
    it('should test itself... I\'ll figure this stuff out later...', function() {      
    });
  });
});

app.js

'use strict';
(function() {
  var app = angular.module('ngGenieApp', ['ngGenie']);
  
  // Makes this modular if we don't just use the global instance and use it as a module instead  
  app.constant('genie', genie);
  
  app.controller('NgGenieCtrl', function($scope, genie) {

    function addWish(wishDisplay, magicWords, action) {
      if (typeof magicWords === 'function') {
        action = magicWords;
        magicWords = undefined;
      }
      if (magicWords === undefined) {
        magicWords = [];
      }
      if (magicWords.indexOf('[') == 0) {
        magicWords = JSON.parse(magicWords);
      }
      magicWords.push(wishDisplay);
      var wish = genie({
        magicWords: magicWords,
        data: {
          displayText: wishDisplay
        },
        action: action || function() {
          console.log('Your "' + this.data.displayText + '" wish is my command!');
        }
      });
    }

    addWish('ng-GenieJS on GitHub', function() {
      window.open('http://www.github.com/kentcdodds/ng-genie', '_blank');
    });
    addWish('GenieJS Demo', function() {
      window.open('http://kentcdodds.github.io/genie', '_blank');
    });
    addWish('GenieJS Tests', function() {
      window.open('http://kentcdodds.github.io/genie/test', '_blank');
    });
    addWish('Create new post');
    addWish('Edit profile pic');
    addWish('Update contact info');
    addWish('Delete account');
    addWish('Block user');
    addWish('Friend user');
    addWish('Find friends');
    addWish('Wow!', ['Alert', 'Pop up'], function() {
      alert('Wow!');
    });

  });
  
})();

ngGenie.js
/**
 * ngGenie.js @license
 * (c) 2013 Kent C. Dodds
 * ngGenie.js may be freely distributed under the MIT license.
 * http://www.github.com/kentcdodds/ng-genie
 * See README.md
 */

angular.module('ngGenie', []).directive('ngLamp', function(genie, $timeout, $document) {
  return {
    replace: true,
    template: function(el, attr) {
      var ngShow = ' ng-show="ngGenieVisible"';
      if (attr.rubClass) {
        ngShow = '';
      }
      return ['<div class="genie-container"' + ngShow + '>',
        '<input type="text" ng-model="genieInput" />',
        '<div class="genie-wishes">',
          '<div class="genie-wish" ' +
            'ng-repeat="wish in matchingWishes" ' +
            'ng-class="{focused: focusedWish == wish}" ' +
            'ng-click="makeWish(wish)" ' +
            'ng-mouseenter="focusOnWish(wish, false)">',
          '{{wish.data.displayText || wish.magicWords[0]}}',
        '</div></div></div>'].join('');
    },
    scope: {
      rubClass: '@',
      rubShortcut: '@',
      rubModifier: '@',
      rubEventType: '@'
    },
    link: function(scope, el, attr) {
      var inputEl = angular.element(el.children()[0]);
      var genieOptionContainer = angular.element(el.children()[1]);
      if (!scope.rubShortcut) {
        scope.rubShortcut = "32";
        scope.rubModifier = "ctrlKey";
      }
      var rubShortcut = parseInt(scope.rubShortcut, 10);
      if (isNaN(rubShortcut)) {
        rubShortcut = scope.rubShortcut[0].charCodeAt(0);
      }

      scope.ngGenieVisible = false;

      // Wish focus
      scope.focusOnWish = function(wishElement, autoScroll) {
        scope.focusedWish = wishElement;
        if (autoScroll) {
          scrollToWish(scope.matchingWishes.indexOf(wishElement));
        }
      };

      function scrollToWish(index) {
        var containerHeight = genieOptionContainer[0].offsetHeight || genieOptionContainer[0].clientHeight;
        var focusedWishElement = genieOptionContainer.children()[index];
        var containerTop = genieOptionContainer[0].scrollTop;
        var containerBottom = containerTop + containerHeight;
        var focusedWishTop = 0;
        angular.forEach(genieOptionContainer.children(), function(child, childIndex) {
          if (childIndex < index) {
            focusedWishTop += child.offsetHeight || child.clientHeight;
          }
        });
        var focusedWishBottom = focusedWishTop + focusedWishElement.offsetHeight || focusedWishElement.clientHeight;
        if (containerBottom < focusedWishBottom) {
          genieOptionContainer[0].scrollTop = focusedWishBottom - containerHeight;
        } else if (containerTop > focusedWishTop) {
          genieOptionContainer[0].scrollTop = focusedWishTop;
        }
      }

      // Document events
      $document.bind('click', function(event) {
        if (el.find(event.srcElement).length < 1) {
          scope.$apply(function() {
            scope.ngGenieVisible = false;
          });
        }
      });

      $document.bind(scope.rubEventType || 'keydown', function(event) {
        if (event.keyCode === rubShortcut) {
          event.preventDefault();
          if (scope.rubModifier) {
            if (event[scope.rubModifier]) {
              scope.ngGenieVisible = !scope.ngGenieVisible;
            }
          } else {
            scope.ngGenieVisible = !scope.ngGenieVisible;
          }
        }
      });

      $document.bind('keydown', function(event) {
        if (event.keyCode === 27 && scope.ngGenieVisible) {
          event.preventDefault();
          scope.ngGenieVisible = false;
        }
      });

      // Input events
      inputEl.bind('keydown', (function() {
        var changeSelection = function(change, event) {
          if (scope.matchingWishes && change) {
            if (event) {
              event.preventDefault();
            }
            var index = scope.matchingWishes.indexOf(scope.focusedWish);
            var newIndex = index + change;
            if (newIndex < 0) {
              newIndex = newIndex + scope.matchingWishes.length;
            } else if (newIndex >= scope.matchingWishes.length) {
              newIndex = newIndex - scope.matchingWishes.length;
            }
            scope.$apply(function() {
              scope.focusOnWish(scope.matchingWishes[newIndex], true);
            });
          }
        }
        return function keydownHandler(event) {
          var change = 0;
          switch(event.keyCode) {
            case 38:
              change = -1;
              break;
            case 40:
              change = 1;
              break;
          }
          if (event.shiftKey) {
            change *= 5;
          }
          changeSelection(change, event);
        }
      })());

      // Making a wish
      scope.makeWish = function(wish) {
        genie.makeWish(wish, scope.genieInput);
        updateMatchingWishes(scope.genieInput);
        scope.ngGenieVisible = false;
      }

      el.bind('keyup', function(event) {
        if (event.keyCode === 13 && scope.focusedWish) {
          genie.makeWish(scope.focusedWish, scope.genieInput);
          scope.$apply(function() {
            updateMatchingWishes(scope.genieInput);
            scope.ngGenieVisible = false;
          });
        }
      });

      // Updating list of wishes
      function updateMatchingWishes(magicWord) {
        if (magicWord) {
          if (magicWord.indexOf('\'') === 0) {
            magicWord = magicWord.substring(1);
          }
          scope.matchingWishes = genie.getMatchingWishes(magicWord);
          if (scope.matchingWishes.length > 0) {
            scope.focusedWish = scope.matchingWishes[0];
          } else {
            scope.focusedWish = null;
          }
        } else {
          scope.matchingWishes = null;
          scope.focusedWish = null;
        }
      }

      scope.$watch('ngGenieVisible', function(newVal) {
        if (newVal) {
          el.addClass(scope.rubClass);
          // Needs to be visible before it can be selected
          $timeout(function() {
            inputEl[0].select();
          }, 25);
        } else {
          el.removeClass(scope.rubClass);
          inputEl[0].blur();
        }
      });

      scope.$watch('genieInput', function(newVal) {
        updateMatchingWishes(newVal);
      });

    }
  }
});


uxGenie.js
/**
 * uxGenie.js @license
 * (c) 2014 Kent C. Dodds
 * uxGenie.js may be freely distributed under the MIT license.
 * http://www.github.com/kentcdodds/ux-genie
 * See README.md
 */
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    define(['./genie'], factory);
  } else if (typeof module !== 'undefined' && module.exports) {
    module.exports = factory(require('geniejs'));
  } else {
    root.genieUx = factory(genie);
  }
}(this, function(genie) {

  var uxGenie = angular.module('uxGenie', []);

  uxGenie.constant('genie', genie);

  uxGenie.directive('uxLamp', ['genie', '$timeout', '$document', function(genie, $timeout, $document) {
    var states = {
      userEntry: 'userentry',
      subContext: 'subcontext'
    };
    return {
      replace: true,
      transclude: true,
      template: function(el, attr) {
        var ngShow = !attr.rubClass ? ' ng-show="lampVisible"' : '';
        var template = '<div class="genie-lamp-container"' + ngShow + ' ng-transclude></div>';
        if (!attr.uxLamp) {
          template = ['<div class="genie-lamp-container"' + ngShow + '>',
            '<input type="text" ng-model="uxLamp.input" class="lamp-input input form-control" />',
            '<div ng-show="uxLamp.matchingWishes.length > 0" class="lamp-wishes-container">',
              '<div class="lamp-wish wish-{{wish.id}}" ' +
                'ng-repeat="wish in uxLamp.matchingWishes" ' +
                'ng-class="{focused: uxLamp.focusedWish == wish}" ' +
                'ng-click="uxLamp.makeWish(wish)" ' +
                'ng-mouseenter="uxLamp.focusOnWish(wish, false)">',
                  '<span class="wish-icon" ng-class="{\'has-img\': wish.data.uxGenie.imgIcon, \'has-i\': wish.data.uxGenie.iIcon}">',
                    '<img class="wish-img-icon" ng-if="wish.data.uxGenie.imgIcon" ng-src="{{wish.data.uxGenie.imgIcon}}">',
                    '<i class="wish-i-icon {{wish.data.uxGenie.iIcon}}" ng-if="wish.data.uxGenie.iIcon"></i>',
                  '</span>',
                  '<span class="wish-display-text">{{wish.data.uxGenie.displayText || wish.magicWords[0]}}</span>',
                '</div>',
              '</div>',
            '</div>'].join('');
        }
        return template;
      },
      scope: {
        uxLamp: '=?',
        lampVisible: '=?',
        rubClass: '@',
        rubShortcut: '@',
        rubModifier: '@',
        rubEventType: '@',
        wishCallback: '&?',
        localStorage: '@',
        firebase: '@'
      },
      link: function(scope, el) {
        scope.uxLamp = scope.uxLamp || {};
        scope.uxLamp.input = '';
        scope.uxLamp.state = states.userEntry;
        scope.lampVisible = false;

        var mathResultId = 'ux-genie-math-result';
        var startTextForSubContext = null;
        var preSubContextContext = null;

        function findFirstChild(el, className) {
          var container = null;
          if (el.hasClass(className)) {
            container = el;
          } else {
            var children = el.children();
            for (var i = 0; i < children.length; i++) {
              container = findFirstChild(angular.element(children[i]), className);
              if (container) {
                break;
              }
            }
          }
          return container;
        }

        var inputEl = findFirstChild(el, 'lamp-input');
        var lampWishesContainer = findFirstChild(el, 'lamp-wishes-container');

        var rubShortcut = scope.rubShortcut || '32';
        var rubModifier = scope.rubModifier || 'ctrlKey';

        rubShortcut = parseInt(rubShortcut, 10);
        if (isNaN(rubShortcut)) {
          rubShortcut = rubShortcut[0].charCodeAt(0);
        }


        /*
         * Setup persistance
         */
        if (scope.firebase && scope.localStorage) {
          throw new Error('ux-lamp cannot have both firebase and local-storage attributes. Choose one or the other.');
        }

        var firebaseRef = null;
        if (scope.firebase) {
          if  (typeof Firebase === 'function') {
            firebaseRef = new Firebase(scope.firebase);
          } else {
            throw new Error('ux-lamp cannot use the given firebase url without the "Firebase" global variable.');
          }
        }

        function saveToLocalStorage(key, words) {
          if (key && localStorage) {
            var json = JSON.stringify(words);
            localStorage.setItem(key, json);
          }
        }

        function saveToFirebase(ref, words) {
          if (ref) {
            ref.set(words);
          }
        }

        function saveGenie () {
          var words = genie.options().enteredMagicWords;
          saveToLocalStorage(scope.localStorage, words);
          saveToFirebase(firebaseRef, words);
        }

        if (firebaseRef) {
          firebaseRef.on('value', function(snapshot) {
            genie.options({
              enteredMagicWords: snapshot.val()
            });
          });
        } else if (scope.localStorage && localStorage) {
          genie.options({
            enteredMagicWords: JSON.parse(localStorage.getItem(scope.localStorage))
          });
        }


        /*
         * Helpers
         */
        function safeApply(fn) {
          var phase = scope.$root.$$phase;
          if(phase == '$apply' || phase == '$digest') {
            scope.$eval(fn);
          }
          else {
            scope.$apply(fn);
          }
        }

        function toggleVisibility(state) {
          safeApply(function() {
            if (typeof state === 'boolean') {
              scope.lampVisible = !state;
            }
            scope.lampVisible = !scope.lampVisible;
          });
        }
        function getElementOuterHeight(element, includeMargin) {
          var cs = document.defaultView.getComputedStyle(element, '');

          var height = parseInt(cs.getPropertyValue('height'));
          var mTop = includeMargin ? parseInt(cs.getPropertyValue('margin-top')) : 0;
          var mBottom = includeMargin ? parseInt(cs.getPropertyValue('margin-bottom')) : 0;
          return height + mTop + mBottom;
        }

        function scrollToWish(index) {
          var containerEl = lampWishesContainer[0];
          var containerHeight = getElementOuterHeight(containerEl);
          var focusedWishElement = lampWishesContainer.children()[index];
          var containerTop = containerEl.scrollTop;
          var containerBottom = containerTop + containerHeight;
          var focusedWishTop = 0;
          var wishElements = lampWishesContainer.children();
          for (var i = 0; i < wishElements.length; i++) {
            if (i >= index) break;
            focusedWishTop += getElementOuterHeight(wishElements[i], true);
          }
          var focusedWishBottom = focusedWishTop + getElementOuterHeight(focusedWishElement, true);
          if (containerBottom < focusedWishBottom) {
            containerEl.scrollTop = focusedWishBottom - containerHeight;
          } else if (containerTop > focusedWishTop) {
            containerEl.scrollTop = focusedWishTop;
          }
        }

        /*
         * Wish focus
         */
        scope.uxLamp.focusOnWish = function(wishElement, autoScroll) {
          scope.uxLamp.focusedWish = wishElement;
          if (scope.uxLamp.focusedWish && autoScroll) {
            scrollToWish(scope.uxLamp.matchingWishes.indexOf(wishElement));
          }
        };

        /*
         * Document events
         */
        $document.bind('click', function(event) {
          if (!scope.lampVisible) return;
          // If it's not part of the lamp, then make the lamp invisible.
          var parent = event.srcElement || event.target;
          while (parent) {
            if (parent === el[0]) {
              return;
            }
            parent = parent.parentNode;
          }
          toggleVisibility(false);
        });

        $document.bind(scope.rubEventType || 'keydown', function(event) {
          if (event.keyCode === rubShortcut) {
            if (rubModifier) {
              if (event[rubModifier]) {
                event.preventDefault();
                toggleVisibility();
              }
            } else {
              event.preventDefault();
              toggleVisibility();
            }
          }
        });

        $document.bind('keydown', function(event) {
          if (event.keyCode === 27 && scope.lampVisible) {
            event.preventDefault();
            toggleVisibility(false);
          }
        });

        /*
         * Input events
         */
        function changeSelection(change, event) {
          var wishes = scope.uxLamp.matchingWishes;
          if (wishes && change) {
            if (event) {
              event.preventDefault();
            }
            var index = wishes.indexOf(scope.uxLamp.focusedWish);
            var newIndex = index + change;
            var totalWishes = wishes.length;
            if (newIndex < 0) {
              newIndex = newIndex + totalWishes;
            } else if (newIndex >= totalWishes) {
              newIndex = newIndex - totalWishes;
            }
            safeApply(function() {
              scope.uxLamp.focusOnWish(wishes[newIndex], true);
            });
          }
        }
        
        inputEl.bind('keydown', (function() {
          return function keydownHandler(event) {
            var change = 0;
            switch(event.keyCode) {
              case 9:
                event.preventDefault();
                var focusedWish = scope.uxLamp.focusedWish;
                if (_isSubContextWish(focusedWish)) {
                  _setSubContextState(focusedWish);
                }
                break;
              case 38:
                change = -1;
                break;
              case 40:
                change = 1;
                break;
            }
            if (event.shiftKey) {
              change *= 5;
            }
            changeSelection(change, event);
          }
        })());

        function _setSubContextState(wish) {
          if (scope.uxLamp.state !== states.subContext) {
            scope.uxLamp.state = states.subContext;
            startTextForSubContext = wish.magicWords[0] + ' ';
            if (wish.data && wish.data.uxGenie && wish.data.uxGenie.displayText) {
              startTextForSubContext = wish.data.uxGenie.displayText;
            }
            preSubContextContext = genie.context();
            genie.context(wish.data.uxGenie.subContext);
            safeApply(function() {
              scope.uxLamp.input = startTextForSubContext;
            });
          }
        }

        function _exitSubContext() {
          genie.context(preSubContextContext);
          scope.uxLamp.state = states.userEntry;
          startTextForSubContext = null;
          preSubContextContext = null;
        }

        /*
         * Making a wish
         */
        scope.uxLamp.makeWish = function(wish) {
          var makeWish = true;
          var magicWord = scope.uxLamp.input;
          if (magicWord.indexOf('\'') === 0) {
            magicWord = magicWord.substring(1);
          }
          if (scope.uxLamp.state === states.subContext) {
            magicWord = magicWord.substring(startTextForSubContext.length);
          }
          var makeInvisible = true;
          if (wish.id === mathResultId) {
            makeWish = false;
          }

          if (_isSubContextWish(wish)) {
            // Make the wish before the context changes.
            genie.makeWish(wish, magicWord);
            saveGenie();
            _setSubContextState(wish);
            makeInvisible = false;
            makeWish = false;
          }

          if (makeWish) {
            wish = genie.makeWish(wish, magicWord);
            saveGenie();
          }

          scope.wishCallback({
            wish: wish,
            magicWord: magicWord
          });
          if (makeInvisible) {
            toggleVisibility(false);
          }
        };

        el.bind('keyup', function(event) {
          if (event.keyCode === 13 && scope.uxLamp.focusedWish) {
            scope.uxLamp.makeWish(scope.uxLamp.focusedWish);
          }
        });

        /*
         * Updating list of wishes
         */
        function updateMatchingWishes(magicWord) {
          if (magicWord) {
            if (magicWord.indexOf('\'') === 0) {
              magicWord = magicWord.substring(1);
            }
            scope.uxLamp.matchingWishes = genie.getMatchingWishes(magicWord);
            if (scope.uxLamp.matchingWishes.length > 0) {
              scope.uxLamp.focusedWish = scope.uxLamp.matchingWishes[0];
            } else {
              scope.uxLamp.focusedWish = null;
            }
          } else {
            scope.uxLamp.matchingWishes = null;
            scope.uxLamp.focusedWish = null;
          }
        }

        function handleInputChange(newVal) {
          if (scope.uxLamp.state === states.subContext) {
            if (newVal.indexOf(startTextForSubContext.trim()) === 0) {
              newVal = newVal.substring(startTextForSubContext.length);
            } else {
              _exitSubContext();
            }
          }
          updateMatchingWishes(newVal);
          var firstWish = null;
          var firstWishDisplay = null;
          if (scope.uxLamp.matchingWishes && scope.uxLamp.matchingWishes.length > 0) {
            firstWish = scope.uxLamp.matchingWishes[0];
            firstWishDisplay = firstWish.magicWords[0];
            if (firstWish.data && firstWish.data.uxGenie && firstWish.data.uxGenie.displayText) {
              firstWishDisplay = firstWish.data.uxGenie.displayText;
            }
          }

          if (firstWish && scope.uxLamp.matchingWishes.length === 1 &&
            _isSubContextWish(firstWish) && firstWishDisplay === newVal) {
            _setSubContextState(firstWish);
          }

          if (scope.uxLamp.state !== states.subContext) {
            var result = _evaluateMath(newVal || '');
            if (angular.isNumber(result)) {
              scope.uxLamp.matchingWishes = scope.uxLamp.matchingWishes || [];
              scope.uxLamp.matchingWishes.unshift({
                id: mathResultId,
                data: {
                  uxGenie: {
                    displayText: newVal + ' = ' + result
                  }
                }
              });
              scope.uxLamp.focusedWish = scope.uxLamp.matchingWishes[0];
            }
          }
        }

        scope.$watch('lampVisible', function(lampIsVisible) {
          if (lampIsVisible) {
            handleInputChange(scope.uxLamp.input);
            if ($document[0].activeElement !== inputEl[0]) {
              if (scope.rubClass) {
                el.addClass(scope.rubClass);
                // Needs to be visible before it can be selected
                $timeout(function() {
                  inputEl[0].select();
                });
              } else {
                inputEl[0].select();
              }
            }
          } else {
            if (scope.rubClass) {
              el.removeClass(scope.rubClass);
            }
            inputEl[0].blur();
          }
        });

        function _isSubContextWish(wish) {
          return !!wish && !!wish.data && !!wish.data.uxGenie && !!wish.data.uxGenie.subContext;
        }

        function _evaluateMath(expression) {
          var mathRegex = /(?:[a-z$_][a-z0-9$_]*)|(?:[;={}\[\]"'!&<>^\\?:])/ig;
          var valid = true;

          expression = expression.replace(mathRegex, function(match) {
            if (Math.hasOwnProperty(match)) {
              return 'Math.' + match;
            } else {
              valid = false;
            }
          });

          if (!valid) {
            return false;
          } else {
            try {
              return eval(expression);
            } catch (e) {
              return false;
            }
          }
        }

        scope.$watch('uxLamp.input', handleInputChange);
      }
    }
  }]);

  uxGenie.directive('genieWish', ['genie', function(genie) {
    return {
      scope: true,
      link: function(scope, el, attrs) {
        var id = attrs.wishId;
        var context = attrs.wishContext ? attrs.wishContext.split(',') : null;
        var data = attrs.wishData || {};
        var uxGenieData = data.uxGenie = data.uxGenie || {};

        uxGenieData.element = el[0];
        uxGenieData.event = attrs.wishEvent || uxGenieData.event || 'click';
        uxGenieData.iIcon = attrs.wishIIcon;
        uxGenieData.imgIcon = attrs.wishImgIcon;

        var action = function(wish) {
          var modifiers = [];
          if (attrs.eventModifiers) {
            modifiers = attrs.eventModifiers.split(',');
          }
          var event = new MouseEvent(wish.data.uxGenie.event, {
            view: window,
            bubbles: true,
            cancelable: true,
            ctrlKey: modifiers.indexOf('ctrlKey') > -1,
            altKey: modifiers.indexOf('altKey') > -1,
            shiftKey: modifiers.indexOf('shiftKey') > -1,
            metaKey: modifiers.indexOf('metaKey') > -1
          });
          wish.data.uxGenie.element.dispatchEvent(event);
        };

        // get magic words
        var magicWords = null;
        ['genieWish', 'name', 'id'].every(function(attrName) {
          magicWords = attrs[attrName];
          return !magicWords;
        });
        magicWords = magicWords || el.text();
        magicWords = magicWords.replace(/\\,/g, '{{COMMA}}');
        if (magicWords) {
          magicWords = magicWords.split(',');
        } else {
          throw new Error('Thrown by the genie-wish directive: All genie-wish elements must have a magic-words, id, or name attribute.');
        }
        for (var i = 0; i < magicWords.length; i++) {
          magicWords[i] = magicWords[i].replace(/\{\{COMMA\}\}/g, ',');
        }

        var wishRegistered = false;
        attrs.$observe('ignoreWish', function(newVal) {
          if (newVal !== 'true' && !wishRegistered) {
            genie({
              id: id,
              magicWords: magicWords,
              context: context,
              action: action,
              data: data
            });
            wishRegistered = true;
          }
        });
      }
    }
  }]);
}));